src/java.base/share/classes/jdk/internal/ref/CleanerImpl.java

```diff
@@ -152,6 +169,33 @@ public void run() {
         }
     }
 
+    @Override
+    public synchronized void beforeCheckpoint(Context<? extends Resource> context) throws Exception {
+        // We block the cleaner thread to prevent race conditions between this
+        // thread and checkpointing thread invoking clean().
+        // When the cleanup starts in cleaner thread the checkpoint will skip
+        // it, but without waiting for the cleanup to finish (which might be
+        // critical for the checkpoint, e.g. closing FDs).
+        // The limitation is that code performing C/R must not wait on any task
+        // completed by the cleaner.
+        blockForCheckpoint = true;
+        thread.interrupt();
```

Comment:
Why it has to be interrupt and not notify(), for example?

Reply:
The interrupt wakes up cleaner in `queue.remove()` (line 161) in case it's blocking for next task.

Reply:
OK, thanks. But what if the interrupt arrives to the thread during a cleaning while that runs, it may interrupt the cleaning, right?

Reply:
Hmm, this is a problem. ReferenceQueue may have a JDK-internal interface via SharedSecretes to interrupt the wait. Or the interface to remove entries without any wait.